docs/adr/2023-09-08_026-network-resilience.md

```diff
@@ -0,0 +1,68 @@
+* We have built our `Network` infrastructure over the concept of relatively independent layers, each implementing a similar interface with different kind of messages, to `broadcast` messages to all peers and be notified of incoming messages through a `callback`.
+  * This pipes-like abstraction allows us to compose our network stack like:
+
+    ```
+     withAuthentication (contramap Authentication tracer) signingKey otherParties $
+      withHeartbeat nodeId connectionMessages $
+        withOuroborosNetwork (contramap Network tracer) localhost peers
+    ```
+
+  * This has the nice property that we can basically swap the lower layers should we need to, for example to use [UDP](https://github.com/input-output-hk/hydra/blob/abailly-iohk/multi-node-udp/hydra-node/src/Hydra/Network/UDP.hs), or add other layers for example to identify [Multiple Heads](https://github.com/input-output-hk/hydra/blob/abailly-iohk/multi-node-udp/hydra-node/src/Hydra/Network/MultiHead.hs#L26)
+
+## Decision
+
+* We implement our own message tracking and resending logic as a standalone `Network` layer
+* That layer consumes and produces `Authenticate`d messages in order to identify the source of messages
+* It uses a vector of monotonically increasing _sequence numbers_ associated with each party (including itself) to track what are the last messages from each party and to ensure FIFO delivery of messages
+* This _vector_ is also used to identify peers which are lagging behind and to resend the missing messages, or to drop messages which have already been received
+* This _vector_ is also piggybacked by the ping messages of the heartbeat mechanism so that a node informs its peers of its own situation even if it would have, otherwise, no message to sent to them
+* Sending a ping message does not influence the _vector_ of the sender
+* Any message received which index does not match what the peer expects is dropped
+* Messages deemed not received by a peer a re-sent
+
+## Consequences
+
+* We keep our existing `Network` interface hence all messages will be resent to all peers
+  * This could be later optimized either by providing a smarter interface with a `send :: Peer -> msg -> m ()` unicast function, or by adding a layer with filtering capabilities, or both
+* We need to ensure messages are not kept forever when resending, eg. that the pending messages list is garbage collected
+* We need to refactor our `Heartbeat` layer to decouple the 2 capabilities it embeds, namely sending periodical heartbeats to peers, and providing listeners with information about a peer connection status, so that the `Reliability` layer can actually rely on those heartbeats to get regular update of peers' knowledge even when the actual Head is idle.
```

Comment:
Related to my question about interaction here. Does the `Reliability` layer (I called it `Resending` before) depend on the `PeerDisconnected` kind of callback messages yielded by the `Heartbeat` layer?

Reply:
No. It relies only on the messages vector clock sent by a peer. Heartbeat messages can be helpful to piggy-back such information in the absence of applicative messages.